designs/NuGetExtensibility/INuGetProjectServices.md

```diff
@@ -0,0 +1,99 @@
+
+See [NuGet extensibility services](NuGetExtensibilityServices.md) for more details.
+
+## Who are the customers
+
+Visual Studio component and extension developers.
+
+## Goals
+
+* Can be called on any thread (ideally background thread, but work if called on UI thread).
+* Quickly return incomplete task to unblock thread.
+* Works with Visual Studio Codespaces, and hopefully Live Share.
+
+## Non-Goals
+
+* This will not be a 1:1 replacement for any of NuGet's existing `IVs` interfaces.
+
+## Solution
+
+```cs
+public interface INuGetProjectServices
+{
+    /// <summary>Gets the list of packages installed in a project.</summmry>
+    /// <param name="project">Path and filename to project file.</param>
+    /// <param name="options">Options to use. If null, uses GetInstalledPakcagesOptions.Default.</param>
+    /// <param name="cancellationToken">Cancellation token.</param>
+    /// <returns>
+    /// The list of packages in the project.
+    /// If the project does not exist in the solution, throws ArgumentException.
+    /// If the project is not ready, throws NuGetProjectNotReadyException.
+    /// </return>
+    Task<IReadOnlyList<NuGetPackage>> GetInstalledPackagesAsync(string project, GetInstalledPackagesOptions options, CancellationToken cancellationToken);
+}
+
+public sealed class NuGetPackage
+{
+    /// <summary>The package id.</summary>
+    string Id { get; }
+
+    /// <summary>The package version</summary>
+    /// <remarks>
+    /// If the project uses packages.config, this will be the installed package version.
+    /// If the project uses PackageReference, this would ideally be the resolved package version, but may be the requested package version.
+    /// </remarks>
+    string Version { get; }
+
+    internal NuGetPackage(string id, string version);
+}
+
+public sealed class GetInstalledPackagesOptions
+{
+    public static GetInstalledPackageOptions Default { get; }
+
+    // in the future may include options such as WaitForNomination, WaitForRestore.
+}
+
+public sealed class NuGetProjectNotReadyException : InvalidOperationException
+{
+}
+
+```
+
+## Future Work
+
+* APIs for:
+  * InstallPackageAsync
+  * UninstallPackageAsync
+
+## Open Questions
+
+* How to handle errors
+  * CPS projects (SDK-style .NET projects) work by: On project load, CPS sends NuGet a "nomination" with project/package information. Therefore, NuGet can't return packages in GetPackagesAsync until CPS nomination complete. With IVsPackageInstallerServices, we throw InvalidOperationException if we haven't got nomination info. Given that the new API needs to work over remoting scenarios, how should it be done?  There are multiple options
+    1. Throw exception (task in faulted state). This might appear as a RemotingException to callers, so maybe hard to distinguish between different types of exceptions?
```

Comment:
The remote invocation exception has a code. So we can basically just say Code `1` means, this, code `3` means that.